python-package/lightgbm/basic.py

```diff
@@ -1111,13 +1183,155 @@ def __init__(self, data, label=None, reference=None,
         except AttributeError:
             pass
 
+    def create_sample_indices(self, total_nrow: int) -> np.ndarray:
+        """Get an array of randomly chosen indices from this ``Dataset``.
+
+        Indices are sampled without replacement.
+
+        Parameters
+        ----------
+        total_nrow: int
+            Total number of rows to sample from.
+            If this value is greater than the value of parameter ``bin_construct_sample_cnt``, only ``bin_construct_sample_cnt`` indices will be used.
+            If Dataset has multiple input data, this should be the sum of rows of every file.
+
+        Returns
+        -------
+        indices: numpy array
+            Indices for sampled data.
+        """
+        param_str = param_dict_to_str(self.params)
+        # Note self.params may contain 'bin_construct_sample_cnt' but is None.
+        sample_cnt = self.params.get("bin_construct_sample_cnt") or DEFAULT_BIN_CONSTRUCT_SAMPLE_CNT
+        sample_cnt = min(sample_cnt, total_nrow)
+        indices = np.zeros(sample_cnt, dtype=np.int32)
+        ptr_data, _, _ = c_int_array(indices)
+
+        _safe_call(_LIB.LGBM_SampleIndices(
+            ctypes.c_int(total_nrow),
+            c_str(param_str),
+            ptr_data,
+        ))
+        return indices
+
+    def init_from_ref_dataset(self, total_nrow: int, ref_dataset: 'Dataset') -> 'Dataset':
+        """Create dataset from a reference dataset.
+
+        Parameters
+        ----------
+            total_nrow (int): number of rows expected to add to dataset
+            ref_dataset (Dataset): referance dataset to extract meta from
+
+        Returns
+        -------
+        self : Dataset
+            Constructed Dataset object.
+        """
+        self.handle = ctypes.c_void_p()
+        _safe_call(_LIB.LGBM_DatasetCreateByReference(
+            ref_dataset,
+            ctypes.c_int64(total_nrow),
+            ctypes.byref(self.handle),
+        ))
+        return self
+
+    def init_from_sample(
+        self,
+        sample_data: List[np.ndarray],
+        sample_indices: List[np.ndarray],
+        sample_cnt: int,
+        total_nrow: int,
+    ) -> "Dataset":
+        """Create Dataset from sampled data structures.
+
+        Parameters
+        ----------
+        sample_data:
+            Sample data for each column
+        sample_indices:
+            Sample data row index for each column.
+        sample_cnt:
+            Number of samples.
+        total_nrow:
+            Total number of rows for all input file.
+
+        Returns
+        -------
+        self : Dataset
+            Constructed Dataset object.
+        """
+        ncol = len(sample_indices)
+        assert len(sample_data) == ncol, "#sample data column != #column indices"
+
+        for i in range(ncol):
+            if sample_data[i].dtype != np.double:
+                raise ValueError("sample_data[{}] type {} is not double".format(i, sample_data[i].dtype))
```

Comment:
As of #4143, as part of #4136, all string formatting in this file has been converted to use f-strings, for speed and ease of use.

Can you please change this and any other uses of `.format()` to f-strings?

```suggestion
                raise ValueError(f"sample_data[{i}] type {sample_data[i].dtype} is not double")
```